### ISSUE TITLE
Ensure SSH config file loading is enabled by default in 2.0

### ISSUE THREAD

**Issue by yuv** [fabric/fabric#1324]
# tl;dr

Host's names and port defined in SSH configuration file are ignored, leading to a refused connection.
# Detailed Report

I am very new to fabric and already love it! I tried the simple "uname -a" task example from the homepage with unexpected results when different port numbers are used for the SSH connection.  I use *buntu 14.04LTS, latest updates, on both the client (Xubuntu) and the server (Ubuntu server).
## SSH configuration

I am a proficient SSH user. My **`~/.ssh/config`** file has the following entry to connect to my home server _mango_ (on a laptop that can connect to _mango_ locally (here) and via a VPN (there):

```
Host here
  Hostname mango
  Port 22

Host there
  Hostname mango
  Port 2222
```
## Direct Execution

`ssh -t mango "uname -s"`, `ssh -t here "uname -s"`,`ssh -t there "uname -s"`, all yield the expected result.
## Fabric

I use this simple **`./fab1.py`**:

``` python
#!/usr/bin/env python

from fabric.api import run

def host_type():
  run ('uname -s')
```
### Successful (Plain Vanilla) Invocation:

`fab --show=debug -f ./fab1.py -i ~/.ssh/id_rsa --ssh-config-path=~/.ssh/config -H mango host_type`
### Failed Invocation (Hostname Only) with Traceback:

`fab --show=debug -f ./fab1.py -i ~/.ssh/id_rsa --ssh-config-path=~/.ssh/config -H here host_type`

```
Using fabfile '/home/yuv/src/fabric_tasks/fab1.py'
Commands to run: host_type
Parallel tasks now using pool size of 1
[here] Executing task 'host_type'
[here] run: /bin/bash -l -c "uname -s"
Timed out trying to connect to here (attempt 1 of 1), giving up)

Fatal error: Low level socket error connecting to host here on port 22: Connection refused (tried 1 time)

Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 384, in execute
    multiprocessing
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 274, in _execute
    return task.run(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 174, in run
    return self.wrapped(*args, **kwargs)
  File "/home/yuv/src/fabric_tasks/fab1.py", line 6, in host_type
    run ('uname -s')
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 647, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/fabric/operations.py", line 1054, in run
    shell_escape=shell_escape)
  File "/usr/local/lib/python2.7/dist-packages/fabric/operations.py", line 921, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/usr/local/lib/python2.7/dist-packages/fabric/state.py", line 397, in default_channel
    chan = _open_session()
  File "/usr/local/lib/python2.7/dist-packages/fabric/state.py", line 389, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 159, in __getitem__
    self.connect(key)
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 151, in connect
    user, host, port, cache=self, seek_gateway=seek_gateway)
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 573, in connect
    raise NetworkError(msg, e)
NetworkError: Low level socket error connecting to host here on port 22: Connection refused (tried 1 time)


Aborting.
Low level socket error connecting to host here on port 22: Connection refused (tried 1 time)

Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 384, in execute
    multiprocessing
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 274, in _execute
    return task.run(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 174, in run
    return self.wrapped(*args, **kwargs)
  File "/home/yuv/src/fabric_tasks/fab1.py", line 6, in host_type
    run ('uname -s')
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 647, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/fabric/operations.py", line 1054, in run
    shell_escape=shell_escape)
  File "/usr/local/lib/python2.7/dist-packages/fabric/operations.py", line 921, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/usr/local/lib/python2.7/dist-packages/fabric/state.py", line 397, in default_channel
    chan = _open_session()
  File "/usr/local/lib/python2.7/dist-packages/fabric/state.py", line 389, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 159, in __getitem__
    self.connect(key)
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 151, in connect
    user, host, port, cache=self, seek_gateway=seek_gateway)
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 573, in connect
    raise NetworkError(msg, e)
NetworkError: Low level socket error connecting to host here on port 22: Connection refused (tried 1 time)
```
### Failed Invocation (Hostname and Port) with Traceback:

`fab --show=debug -f ./fab1.py -i ~/.ssh/id_rsa --ssh-config-path=~/.ssh/config -H there --port=2222 host_type`

```
Using fabfile '/home/yuv/src/fabric_tasks/fab1.py'
Commands to run: host_type
Parallel tasks now using pool size of 1
[there] Executing task 'host_type'
[there] run: /bin/bash -l -c "uname -s"
Timed out trying to connect to there (attempt 1 of 1), giving up)

Fatal error: Low level socket error connecting to host there on port 2222: Connection refused (tried 1 time)

Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 384, in execute
    multiprocessing
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 274, in _execute
    return task.run(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 174, in run
    return self.wrapped(*args, **kwargs)
  File "/home/yuv/src/fabric_tasks/fab1.py", line 6, in host_type
    run ('uname -s')
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 647, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/fabric/operations.py", line 1054, in run
    shell_escape=shell_escape)
  File "/usr/local/lib/python2.7/dist-packages/fabric/operations.py", line 921, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/usr/local/lib/python2.7/dist-packages/fabric/state.py", line 397, in default_channel
    chan = _open_session()
  File "/usr/local/lib/python2.7/dist-packages/fabric/state.py", line 389, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 159, in __getitem__
    self.connect(key)
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 151, in connect
    user, host, port, cache=self, seek_gateway=seek_gateway)
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 573, in connect
    raise NetworkError(msg, e)
NetworkError: Low level socket error connecting to host there on port 2222: Connection refused (tried 1 time)


Aborting.
Low level socket error connecting to host there on port 2222: Connection refused (tried 1 time)

Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 384, in execute
    multiprocessing
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 274, in _execute
    return task.run(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/fabric/tasks.py", line 174, in run
    return self.wrapped(*args, **kwargs)
  File "/home/yuv/src/fabric_tasks/fab1.py", line 6, in host_type
    run ('uname -s')
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 647, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/fabric/operations.py", line 1054, in run
    shell_escape=shell_escape)
  File "/usr/local/lib/python2.7/dist-packages/fabric/operations.py", line 921, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/usr/local/lib/python2.7/dist-packages/fabric/state.py", line 397, in default_channel
    chan = _open_session()
  File "/usr/local/lib/python2.7/dist-packages/fabric/state.py", line 389, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 159, in __getitem__
    self.connect(key)
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 151, in connect
    user, host, port, cache=self, seek_gateway=seek_gateway)
  File "/usr/local/lib/python2.7/dist-packages/fabric/network.py", line 573, in connect
    raise NetworkError(msg, e)
NetworkError: Low level socket error connecting to host there on port 2222: Connection refused (tried 1 time)
```

**Comment by naphthalene** [comment]
I believe you can use SSH config host names by passing in `--set use_ssh_config` from the CLI or setting it via the `settings` context manager:

.. _use-ssh-config:

``use_ssh_config``
------------------

**Default:** ``False``

Set to ``True`` to cause Fabric to load your local SSH config file.

.. versionadded:: 1.4
.. seealso:: :ref:`ssh-config`

**Comment by bitprophet** [comment]
@naphthalene Yea, I think I left this open & assigned to 2.0 as a reminder to have SSH config loading enabled by default (since in 1.0 it's not, and changing it would be backwards incompatible for many users). But I am going to tweak this ticket's subject a bit so that's clear...I was all set to agree with you & close this before I noticed I'd set labels/milestone & the original issue was old! :)

**Comment by naphthalene** [comment]
Awesome, thanks! If there is any work on Fabric 2.0 you'd like to delegate I'm happy to work on it - I've been using it and very happy with what it offers (minus a few hiccups).

**Comment by bitprophet** [comment]
This is implemented in my private v2 branch now (public VERY SOON 😮) - closing!
